tools/project_manager/editproject.php

```diff
@@ -599,6 +603,40 @@ function set_from_post()
             }
         }
 
+        $this->custom_chars = @$_POST['custom_chars'];
+        if($this->custom_chars)
+        {
+            $codepoints = utf8_codepoints_combining($this->custom_chars);
+
+            // all characters must be unique
+            if($codepoints != array_unique($codepoints))
+            {
+                $errors .= _("The set of custom characters must be unique.")."<br>";
+            }
+
+            // all characters must be normalized
+            if(count(get_nonnormalized_codepoints($codepoints)))
+            {
+                $errors .= _("Custom characters must be in NFC normalized form.")."<br>";
+            }
```

Comment:
Requiring characters to be normalized is legitimate, but for those entering combining diacritics, they may not even realize that what they're entering is not normalized. I deliberately added an `ā` both in normalized and non-normalized forms to see if that would trigger the message, and it did. Would it be realistic to normalize anything being added to a project custom character suite?

If all they're doing is copying a few characters from an existing character suite, or possibly from another web page, it should come to them normalized, but I'm thinking back to the few characters we found in project titles and authors' names that were non-normalized combining characters. It may be rare, but it will happen, and we'll likely wind up with confused PMs.

Reply:
We can easily transparently normalize them and indeed that might be simpler for the users. My only worry is that some PM is going to put in a character that doesn't NFC normalize down to what they expect.

Reply:
Either they have to go out and find the normalized form, or they put in what they think is the right character (or character combination) and we normalize it for them. If it doesn't look right, they still have to go out and find the normalized form. As it stands now, it's slightly more work for them, because if they get the normalization error, they have to go out and find a normalized form when they might not have to if we do it for them.

I noticed one character yesterday where the character was vastly different, depending on the font, so it's also possible that any problems in shape might be due to font variations.

Reply:
Characters are now normalized on save.

Reply:
Looks like the normalization is done after the check for duplicates:
<img width="121" alt="image" src="https://user-images.githubusercontent.com/36800239/88353774-67127100-cd24-11ea-891c-0b601a660b7b.png">

Reference custom charsuite in [UTF-8 practice: Tiger Lily and other stories.](https://www.pgdp.org/~cpeel/c.branch/per-project-charsuite/tools/charsuites.php?projectid=projectID5df912d3cae6b#projectID5df912d3cae6b).

Reply:
Yes, it was on save to the DB. I've moved it up to before the validation. This is probably going to make it confusing for the user if they add a normalized letter and a non-normalized letter and then get a report that there is a duplicate. `<shrugs>`

Reply:
If you think it best, we can go back to just issuing the warning that characters in the custom suite have to be normalized, and not force the normalization. I'm trying to decide what I'd prefer as a PM, and I think I'd prefer the script to take care of it, especially if the only way I know to input it is non-normalized, but others might feel differently, and even if I get a duplicate character message.

Did the original version identify the non-normalized character(s)? (Short memory, here -- I don't remember!) If not, that's a potential problem if the PM doesn't have a clue what's normalized and what's not.